cpp/src/arrow/array.cc

```diff
@@ -831,6 +831,202 @@ Status DictionaryArray::Transpose(MemoryPool* pool, const std::shared_ptr<DataTy
+                           root_out_type->ToString(), ": ", msg);
+  }
+
+  void AdjustInputPointer() {
+    if (input_exhausted) {
+      return;
+    }
+    while (true) {
+      // Skip exhausted layout (might be empty layout)
+      while (in_buffer_idx >= in_layouts[in_layout_idx].bit_widths.size()) {
+        in_buffer_idx = 0;
+        ++in_layout_idx;
+        if (in_layout_idx >= in_layouts.size()) {
+          input_exhausted = true;
+          return;
+        }
+      }
+      if (in_layouts[in_layout_idx].bit_widths[in_buffer_idx] > 0) {
+        return;
+      }
+      // Skip always-null input buffers
+      // (e.g. buffer 0 of a null type or buffer 2 of a sparse union)
+      ++in_buffer_idx;
+    }
+  }
+
+  Status CheckInputAvailable() {
+    if (input_exhausted) {
+      return InvalidView("not enough buffers for view type");
+    }
+    return Status::OK();
+  }
+
+  Status CheckInputExhausted() {
+    if (!input_exhausted) {
+      return InvalidView("too many buffers for view type");
+    }
+    return Status::OK();
+  }
+
+  Status CheckInputHasNoDictionaries() {
+    for (const auto& layout : in_layouts) {
+      if (layout.has_dictionary) {
+        return InvalidView("input has dictionary");
+      }
+    }
+    return Status::OK();
+  }
+
+  Status MakeDataView(const std::shared_ptr<Field>& out_field,
+                      std::shared_ptr<ArrayData>* out) {
+    // XXX What about offsets?
+    const auto out_type = out_field->type();
+    const auto out_layout = out_type->layout();
+    if (out_layout.has_dictionary) {
+      return InvalidView("view type requires dictionary");
+    }
+
+    AdjustInputPointer();
+    int64_t out_length = in_data_length;
+    int64_t out_null_count;
+
+    // No type has a purely empty layout
+    DCHECK_GT(out_layout.bit_widths.size(), 0);
+
+    if (out_layout.bit_widths[0] == 0) {
+      // Assuming null type or equivalent.
+      DCHECK_EQ(out_layout.bit_widths.size(), 1);
+      *out = ArrayData::Make(out_type, out_length, {nullptr}, out_length);
+      return Status::OK();
+    }
+
+    std::vector<std::shared_ptr<Buffer>> out_buffers;
+
+    // Process null bitmap
+    DCHECK_EQ(out_layout.bit_widths[0], 1);
+    if (in_buffer_idx == 0) {
+      // Copy input null bitmap
+      RETURN_NOT_OK(CheckInputAvailable());
+      const auto& in_data_item = in_data[in_layout_idx];
+      if (!out_field->nullable() && in_data_item->GetNullCount() != 0) {
+        return InvalidView("nulls in input cannot be viewed as non-nullable");
+      }
+      DCHECK_GT(in_data_item->buffers.size(), in_buffer_idx);
+      out_buffers.push_back(in_data_item->buffers[in_buffer_idx]);
+      out_length = in_data_item->length;
+      out_null_count = in_data_item->null_count;
+      ++in_buffer_idx;
+      AdjustInputPointer();
+    } else {
+      // No null bitmap in input, append no-nulls bitmap
+      out_buffers.push_back(nullptr);
+      out_null_count = 0;
```

Comment:
Does this occur anywhere now?

Reply:
Yes, when viewing structs as structs of different shapes (for example `struct<a: struct<b: int8>, d: int32>` as `struct<a: uint8, b: struct<b: float>>`).